returnn/tf/util/data.py

```diff
@@ -406,9 +409,26 @@ def short_repr(self):
       """
       return "Beam{%r}(%s)" % (self.beam.name, self.size)
 
-  class VariableDim(VirtualDimBase):
+  class PaddedDim(FixedDim):
+    """
+    Represents a dim with variable size, which is flattened with padding (not packed) into the batch.
+    """
+    def __init__(self, dim_tag):
+      """
+      :param DimensionTag dim_tag:
+      """
+      super(BatchInfo.PaddedDim, self).__init__(size=dim_tag.get_dim_value())
+      self.dim_tag = dim_tag
+
+    def short_repr(self):
+      """
+      :rtype: str
+      """
+      return "Padded{%r}" % self.dim_tag.description
+
+  class PackedDim(VirtualDimBase):
     """
-    Represents a dim with variable sizes.
+    Represents a dim with variable sizes, which is packed (un-padded) into the batch.
     Variable w.r.t. other dims (must be per batch entry).
     """
     def __init__(self, sizes, key_axes, dim_tag):
```

Comment:
Yea, this is unclear. But it could be different. `dim_tag` is usually the base dim tag (via `get_same_base()`). `dim_tag.get_same_base().dyn_size` is always without any beam. `dim_tag.dyn_size` could be with beam.

Unfortunately this is already somewhat messy/unclear in `DimensionTag`. Because this was kind of an afterthought, that sizes can be different depending on the beam/batch, but they still correspond to the same dim tag (`==` would be true). And normally other code should resolve different beams.

Reply:
Makes sense! But why don't we change the `dim_tag` attribute to a copy with the correct `dyn_size` whenever we want to change the size for this virtual dim, as opposed to leaving the `dim_tag` as is and changing `sizes`?
Maybe we should make this more clear in `DimensionTag` instead of patching around the issue here? (Perhaps it even introduces more issues if the behavior of the virtual dims is different than the ones of the `DimensionTags`)

Reply:
I just wanted to make sure that the dict `self._packed_dims_by_dim_tag` behaves reasonable and use the "canonical dimension tag" (although, in principle, it should also work with descendants, because they set `__eq__` and `__hash__`). There could be some strange/unexpected `DimensionTag` instances (e.g. referring to tensors which are only valid inside the while loop). It just felt safer that way, because I did not fully trust the correctness of other code...

But yes, maybe it should work. And you are right, it would be cleaner. I will try.